examples/3_simple_storage.py

```diff
@@ -0,0 +1,176 @@
+
+    KVHandler listens to chat messages of the form:
+    `!storage put <namespace> <key> <value> (<revision>)`
+    `!storage get <namespace> <key>`
+    `!storage delete <namespace> <key> (<revision>)`
+    `!storage list`  // list namespaces
+    `!storage list <namespace>`  // list entries in namespace
+    """
+
+    MSG_PREFIX = "!storage"
+
+    async def __call__(self, bot, event):
+        members_type = event.msg.channel.members_type
+        if not event.type == EventType.CHAT:
+            return
+
+        channel = event.msg.channel
+        user = event.msg.sender.username
+
+        # support teams and implicit self teams
+        team = channel.name
+        if members_type == "impteamnative" and channel.name == user:
+            team = "{0},{0}".format(channel.name)
+
+        msg = ""
+        try:
+            msg = event.msg.content.text.body.strip().split(" ")
+        except AttributeError:
+            return
+
+        if len(msg) < 2 or msg[0] != self.MSG_PREFIX:
+            return
+
+        action = msg[1]
+        if action == KVMsg.HELP.value:
+            return await self.handle_help(bot, channel, team, msg, action)
+        if action == KVMsg.LIST.value:
+            return await self.handle_list(bot, channel, team, msg, action)
+        if action == KVMsg.GET.value:
+            return await self.handle_get(bot, channel, team, msg, action)
+        if action == KVMsg.PUT.value:
+            return await self.handle_put(bot, channel, team, msg, action)
+        if action == KVMsg.DELETE.value:
+            return await self.handle_delete(bot, channel, team, msg, action)
+        await bot.chat.send(channel, "invalid !storage command")
+        return
+
+    async def handle_help(self, bot, channel, team, msg, action):
+        if len(msg) == 2:
+            # !storage help
+            send_msg = "Available commands:\
+                    \n`!storage put <namespace> <key> <value> (<revision>)`\
+                    \n`!storage get <namespace> <key>`\
+                    \n`!storage delete <namespace> <key> (<revision>)`\
+                    \n`!storage list  // list namespaces`\
+                    \n`!storage list <namespace>  // list entries in namespace`"
+            await bot.chat.send(channel, send_msg)
+            return
+
+    async def handle_list(self, bot, channel, team, msg, action):
+        if len(msg) == 2:
+            # !storage list
+            res = await bot.kvstore.list_namespaces(team)
+            await bot.chat.send(channel, str(res))
+            return
+        if len(msg) == 3:
+            # !storage list <namespace>
+            namespace = msg[2]
+            res = await bot.kvstore.list_entrykeys(team, namespace)
+            await bot.chat.send(channel, str(res))
+            return
+
+    async def handle_get(self, bot, channel, team, msg, action):
+        if len(msg) == 4:
+            namespace, key = msg[2], msg[3]
+            # !storage get <namespace> <key>
```

Comment:
i wish i'd thought of josh's comment when we were starting this. implicitly, it looks like you can't have spaces in namespaces and entryKeys because we're splitting on them. but that's not at all the case if you're calling into the client directly. whatever. we can ignore it for now.

Reply:
did it~